src/main/scala/is/hail/distributedmatrix/BlockMatrixIsDistributedMatrix.scala

```diff
@@ -0,0 +1,179 @@
+package is.hail.distributedmatrix
+
+import is.hail.utils._
+import org.apache.spark.rdd.RDD
+import org.apache.spark.mllib.linalg._
+import org.apache.spark.mllib.linalg.distributed._
+
+import scala.reflect.ClassTag
+
+object BlockMatrixIsDistributedMatrix extends DistributedMatrix[BlockMatrix] {
+  type M = BlockMatrix
+
+  def cache(m: M): M = m.cache()
+
+  def from(rdd: RDD[Array[Double]]): M =
+    new IndexedRowMatrix(rdd.zipWithIndex().map { case (x, i) => new IndexedRow(i, new DenseVector(x)) })
+      .toBlockMatrixDense()
+  def from(bm: BlockMatrix): M = bm
+  def from(cm: CoordinateMatrix): M = cm.toBlockMatrix()
+
+  def transpose(m: M): M = m.transpose
+  def diagonal(m: M): Array[Double] = toCoordinateMatrix(m)
+    .entries
+    .filter(me => me.i == me.j)
+    .map(me => (me.i, me.value))
+    .collect()
+    .sortBy(_._1)
+    .map(_._2)
+
+  def multiply(l: M, r: M): M = l.multiply(r)
+  def multiply(l: M, r: DenseMatrix): M = {
+    require(l.numCols() == r.numRows,
+      s"incompatible matrix dimensions: ${l.numRows()}x${l.numCols()} and ${r.numRows}x${r.numCols}")
+    val sc = l.blocks.sparkContext
+    val rbc = sc.broadcast(r)
+    val rRowsPerBlock = l.colsPerBlock
+    val rColsPerBlock = l.rowsPerBlock
+    val rRowBlocks = (r.numRows - 1) / rRowsPerBlock + 1
+    val rColBlocks = (r.numCols - 1) / rColsPerBlock + 1
+    val rRowsRemainder = r.numRows % rRowsPerBlock
+    val rColsRemainder = r.numCols % rColsPerBlock
+    val indices = for {
+      i <- 0 until rRowBlocks
+      j <- 0 until rColBlocks
+    } yield (i, j)
+    val rMats = sc.parallelize(indices).map { case (i, j) =>
+      val rRowsInThisBlock = (if (i + 1 == rRowBlocks && rRowsRemainder != 0) rRowsRemainder else rRowsPerBlock)
+      val rColsInThisBlock = (if (i + 1 == rColBlocks && rColsRemainder != 0) rColsRemainder else rColsPerBlock)
+      val a = new Array[Double](rRowsInThisBlock * rColsInThisBlock)
+      for {
+        ii <- 0 until rRowsInThisBlock
+        jj <- 0 until rColsInThisBlock
+      } {
+        a(jj*rRowsInThisBlock + ii) = rbc.value(i*rRowsPerBlock + ii, j*rColsPerBlock + jj)
```

Comment:
1. I'm confused by your response.  The conceptual operation you're doing here is DenseMatrix => BlockMatrix, right?  I'm just suggesting you split this out as a from(DenseMatrix) function in the API.  (I'm amazed Spark doesn't have this, but it appears not to.)